### DISCUSSION TITLE
Disabling Prepared statements

### DISCUSSION THREAD

**Discussion by TheOtherBrian1** [supabase/supabase#28239] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/disabling-prepared-statements-qL8lEL)._

## How to disable prepared statements for Supavisor in transaction mode

Each ORM or library configures prepared statements differently. Here are settings for some common ones. If you don't see yours, make a comment

## Prisma:

add ?pgbouncer=true to end of connection string:

    postgres://[db-user].[project-ref]:[db-password]@aws-0-[aws-region].pooler.supabase.com:6543/[db-name]?pgbouncer=true

## Drizzle:

Add a prepared false flag to the client:

```ts
export const client = postgres(connectionString, { prepare: false })
```

## Node Postgres

[Omit the "name" value in a query definition](https://node-postgres.com/features/queries#prepared-statements):

```ts
const query = {
  name: 'fetch-user', // <--------- DO NOT INCLUDE
  text: 'SELECT * FROM user WHERE id = $1',
  values: [1],
}
```

## Psycopg

set the [prepare\_threshold](https://www.psycopg.org/psycopg3/docs/api/connections.html#psycopg.Connection.prepare_threshold) to `None`.

## asyncpg

Follow the recommendation in the [asyncpg docs](https://magicstack.github.io/asyncpg/current/faq.html#why-am-i-getting-prepared-statement-errors)

> disable automatic use of prepared statements by passing `statement_cache_size=0` to [asyncpg.connect()](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.connect) and [asyncpg.create\_pool()](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.pool.create_pool) (and, obviously, avoid the use of [Connection.prepare()](https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.prepare));

## JDBC

Set [`prepareThreshold`](https://jdbc.postgresql.org/documentation/use/#connection-parameters) to `0`:

    jdbc:postgresql://[POOLER-HOST]:6543/postgres?prepareThreshold=0

## Rust's Deadpool or `tokio-postgres`:

*   Check [GitHub Discussion](https://github.com/bikeshedder/deadpool/issues/340#event-13642472475)

**Comment by mirusky** [comment]
I'm having issues with [GORM](https://gorm.io/index.html) I've setted `PrepareStmt` to false as [docs](https://gorm.io/docs/performance.html#Caches-Prepared-Statement) suggests.

I'll try the 'pgbouncer=true' at end of connection string to see if it takes in.

**Reply by TheOtherBrian1** [reply]
`pgbouncer=true` is how Prisma reconfigures prepared statements, but it would not have an impact on GORM. As the doc you shared pointed out, modifying the [global](https://gorm.io/docs/gorm_config.html#PrepareStmt) and/or [session](https://gorm.io/docs/session.html) configs to `PrepareStmt: false` should be sufficient. 

Are you certain that the error is prepared statement related? 

As an aside, I'm curious to know where you're deploying to: AWS Lambda, Render, Fly.io? Depending on whether the service auto-scales or is predominantly stationary, the direct connection or Supavisor in session mode may be better

**Reply by mirusky** [reply]
My logs show that it is related to the prepared stmt, but even disabling it explicitly continues to give an error. Especially after the app has restarted.
![image](https://github.com/user-attachments/assets/09b54c59-02ce-4e68-b0d1-8f66a63c1e26)

Maybe something related to GORM itself? I'll try to contact them.

**Reply by TheOtherBrian1** [reply]
> I'll try to contact them.

@mirusky, I'd love if you could share their feedback. If we can add it to the list, that could be very helpful for others

**Comment by 2bashbrothers** [comment]
What about postgres.js? I've tried ?pgbouncer=true but no luck
![image](https://github.com/user-attachments/assets/216d69de-3dc1-4ecc-9868-ec5e769cf2fa)

**Reply by 2bashbrothers** [reply]
actually think I just found it:
![image](https://github.com/user-attachments/assets/22a6ed81-99b5-4863-85f3-ab309fa0c4e7)

**Reply by TheOtherBrian1** [reply]
I'll add it to the list. Thanks for the help

**Comment by tedsecretsource** [comment]
What happens if we don't use prepared statements and also don't disable them?

**Reply by TheOtherBrian1** [reply]
@tedsecretsource, I think it may help if I provide a more complete breakdown of what prepared statements are and why they need to be disabled

### What happens when Postgres receives a query

When Postgres receives a query, it first:
- Parses the SQL into a lexical tree
- Replaces query placeholders with their actual arguments
- Selects the algorithms used to retrieve the underlying data

These stages together are known as the planning phase. It is usually very quick (under 10ms, but often sub-ms). You can actually see how long it takes relative to fetching the data (execution phase), by running the following SQL:

```sql
EXPLAIN ANALYZE <your_query>;
```

Even though planning is fast, it still takes time. ORMs like Prisma reduce its overhead by using Prepared Statements (P.S.), cached plans that can be reused, so the database doesn’t replan each time.

### Why prepared statements break with some poolers

The issue is that P.S. exists only within the session/connection they were defined in. If a query refers to a P.S. that doesn't exist in its session or tries to create one that already exists, Postgres will throw an error:

- `ERROR: prepared statement "SOME_NAME" already exists`
- `ERROR: prepared statement "SOME_NAME" does not exist`

Supavisor, when running in transaction mode (port 6543), acts as a Postgres connection load balancer. It dynamically reassigns database connections to clients as needed. So, there is no guarantee that when the query references a P.S., it will be able to find it in the session, causing the above errors.

## What happens if we don't use prepared statements and also don't disable them?

Certain ORMs and drivers will only create prepared statements if explicitly programmed to. For instance, Node Postgres does not make P.S. by default. With these types of libraries, you can use Supavisor in transaction mode without any configurations.

Some ORMs, such as Prisma, will create them in the background by default. Even if you have no desire to use them, you'll still have to manually disable them in your code.
